pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/AbstractApexNode.java

```diff
@@ -77,61 +159,72 @@ public final String getXPathNodeName() {
         return this.getClass().getSimpleName().replaceFirst("^AST", "");
     }
 
-    /**
-     * Note: in this routine, the node has not been added to its parents,
-     * but its children have been populated (except comments).
-     */
-    void closeNode(TextDocument positioner) {
-        // do nothing
-    }
-
     protected void setRegion(TextRegion region) {
         this.region = region;
     }
+    
+    @Override
+    public abstract boolean hasRealLoc();
 
-    @Deprecated
-    @InternalApi
     @Override
-    public T getNode() {
-        return node;
+    public String getDefiningType() {
+        BaseApexClass<?> baseNode = this instanceof BaseApexClass ? (BaseApexClass<?>) this : getFirstParentOfType(BaseApexClass.class);
+        if (baseNode != null) {
+            return baseNode.getQualifiedName().toString();
+        }
+        return null;
     }
 
     @Override
-    public boolean hasRealLoc() {
-        try {
-            Location loc = node.getLoc();
-            return loc != null && Locations.isReal(loc);
-        } catch (UnexpectedCodePathException e) {
-            return false;
-        } catch (IndexOutOfBoundsException | NullPointerException e) {
-            // bug in apex-jorje? happens on some ReferenceExpression nodes
-            return false;
-        }
+    public String getNamespace() {
+        // TypeInfo definingType = getDefiningTypeOrNull();
+        // if (definingType != null) {
+        //     return definingType.getNamespace().toString();
+        // }
+        // TODO(b/243905954)
+        return null;
     }
 
-    private TypeInfo getDefiningTypeOrNull() {
-        try {
-            return node.getDefiningType();
-        } catch (UnsupportedOperationException e) {
-            return null;
+    /** Returns the string value of the {@link LiteralExpression}. */
+    protected static String literalToString(LiteralExpression expr) {
+        if (expr instanceof LiteralExpression.StringVal) {
+            return ((LiteralExpression.StringVal) expr).getValue();
+        } else if (expr instanceof LiteralExpression.NullVal) {
+            return "";
         }
+        return expr.asCodeString();
     }
 
-    @Override
-    public String getDefiningType() {
-        TypeInfo definingType = getDefiningTypeOrNull();
-        if (definingType != null) {
-            return definingType.getApexName();
-        }
-        return null;
+    /**
+      * Normalizes case of primitive type names.
+      *
+      * All other strings are returned unchanged.
+      *
+      * See: https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/langCon_apex_primitives.htm
+      */
+    public static String caseNormalizedTypeIfPrimitive(String name) {
```

Comment:
This is a implementation detail and should not be public - maybe we can keep it package private, so that it can only be called internally by the AST nodes?